docs/develop/dotnet/message-passing.mdx

```diff
@@ -1,261 +1,791 @@
+The code that follows is part of a [working solution](https://github.com/temporalio/samples-dotnet/tree/main/src/MessagePassing).
+:::
+
+Follow these guidelines when writing your message handlers:
+
+<div style={{backgroundColor: '#ffff00',padding: '0px 15px',borderRadius: '5px',border: '1px solid #cccccc',display: 'inline-block'}}>**Check these links.**</div>
+- Message handlers are defined as methods on the Workflow class, using one of the three attributes: [`WorkflowQueryAttribute`](https://dotnet.temporal.io/api/Temporalio.Workflows.WorkflowQueryAttribute.html), [`WorkflowSignalAttribute`](https://dotnet.temporal.io/api/Temporalio.Workflows.WorkflowSignalAttribute.html), and [`WorkflowUpdateAttribute`](https://dotnet.temporal.io/api/Temporalio.Workflows.WorkflowUpdateAttribute.html).
+- The parameters and return values of handlers and the main Workflow function must be [serializable](/dataconversion).
+- Prefer data classes to multiple input parameters.
+  Data class parameters allow you to add fields without changing the calling signature.
 
-- [Develop with Signals](#signals)
-- [Develop with Queries](#queries)
-- [Develop with Updates](#updates)
+### Query handlers {#queries}
 
-## Signals {#signals}
+A [Query](/encyclopedia/workflow-message-passing#sending-queries) is a synchronous operation that retrieves state from a Workflow Execution.
+Define as a method:
+
+```csharp
+[Workflow]
+public class GreetingWorkflow
+{
+    public enum Language
+    {
+        Chinese,
+        English,
+        French,
+        Spanish,
+        Portuguese,
+    }
+
+    public record GetLanguagesInput(bool IncludeUnsupported);
+
+    // ...
+
+    [WorkflowQuery]
+    public IList<Language> GetLanguages(GetLanguagesInput input) =>
+        Enum.GetValues<Language>().
+            Where(language => input.IncludeUnsupported || Greetings.ContainsKey(language)).
+            ToList();
+
+    // ...
+```
 
-**How to develop with Signals using the Temporal .NET SDK**
+Or as a property getter:
 
-A [Signal](/encyclopedia/workflow-message-passing#sending-signals) is a message sent to a running Workflow Execution.
+```csharp
+[Workflow]
+public class GreetingWorkflow
+{
+    public enum Language
+    {
+        Chinese,
+        English,
+        French,
+        Spanish,
+        Portuguese,
+    }
 
-Signals are defined in your code and handled in your Workflow Definition.
-Signals can be sent to Workflow Executions from a Temporal Client or from another Workflow Execution.
+    // ...
 
-### Define Signal {#define-a-signal}
+    [WorkflowQuery]
+    public Language CurrentLanguage { get; private set; } = Language.English;
 
-**How to define a Signal using the Temporal .NET SDK**
+    // ...
+```
 
-A Signal has a name and can have arguments.
+- The Query attribution can accept arguments. <div style={{backgroundColor: '#ffff00',padding: '0px 15px',borderRadius: '5px',border: '1px solid #cccccc',display: 'inline-block'}}>**Example, like... and show?**</div>
```

Comment:
> Example, like... and show?

I didn't see an example of this in Python that I can make the .NET equivalent of. Is there benefit to solving this only in .NET?